src/Build/BackEnd/Node/RarNode.cs

```diff
@@ -0,0 +1,119 @@
+        /// </summary>
+        private const int ClientConnectTimeout = 60000;
+
+        /// <summary>
+        /// Fully qualified name of RarController, used for providing <see cref="IRarController" /> instance to <see cref="RarNode" />
+        /// </summary>
+        const string RarControllerName = "Microsoft.Build.Tasks.ResolveAssemblyReferences.Server.RarController, Microsoft.Build.Tasks.Core";
+
+        public NodeEngineShutdownReason Run(bool nodeReuse, bool lowPriority, out Exception shutdownException)
+        {
+            shutdownException = null;
+            using CancellationTokenSource cts = new CancellationTokenSource();
+            string pipeName = CommunicationsUtilities.GetRarPipeName(nodeReuse, lowPriority);
+            IRarController controller = GetController(pipeName);
+
+            Task<int> rarTask = controller.StartAsync(cts.Token);
+
+            Handshake handshake = NodeProviderOutOfProc.GetHandshake(enableNodeReuse: nodeReuse,
+                                                                     enableLowPriority: lowPriority, specialNode: true);
+            Task<NodeEngineShutdownReason> msBuildShutdown = RunShutdownCheckAsync(handshake, cts.Token);
+
+            // Wait for any of these task to finish:
+            // - rarTask can timeout (default is 15 mins)
+            // - msBuildShutdown ends when it recieves command to shutdown
+            int index = Task.WaitAny(msBuildShutdown, rarTask);
+            cts.Cancel();
+
+            if (index == 0)
+            {
+                // We know that the task completed, so we can get Result without waiting for it.
+                return msBuildShutdown.Result;
+            }
+            else
+            {
+                // RAR task can only exit with connection error or timeout
+                return NodeEngineShutdownReason.ConnectionFailed;
+            }
+        }
+
+        private static IRarController GetController(string pipeName)
+        {
+            Type rarControllerType = Type.GetType(RarControllerName);
+
+            Func<string, int?, int?, int, bool, NamedPipeServerStream> streamFactory = NamedPipeUtil.CreateNamedPipeServer;
+            IRarController controller = (IRarController)Activator.CreateInstance(rarControllerType, pipeName, streamFactory, null);
+
+            ErrorUtilities.VerifyThrow(controller != null, ResourceUtilities.GetResourceString("RarControllerReflectionError"), RarControllerName);
+            return controller;
+        }
+
+
+        public NodeEngineShutdownReason Run(out Exception shutdownException)
+        {
+            return Run(false, false, out shutdownException);
+        }
+
+        private async Task<NodeEngineShutdownReason> RunShutdownCheckAsync(Handshake handshake, CancellationToken cancellationToken = default)
+        {
+            string pipeName = NamedPipeUtil.GetPipeNameOrPath("MSBuild" + Process.GetCurrentProcess().Id);
+
+            while (true)
```

Comment:
It will in most cases not run just once. When MSBuild is connecting to OutOfProc nodes it discovers all running processes and then tries to connect to them. The handshake validation will prevent these unwanted connections. But after that we still have to check if we are receiving correct packet (just to be sure). So most common path will be through the `continue` statement (in long run). But after we get after the handshake, we can be almost sure that this will be shutdown packet

Reply:
Will add comment describing this.